### PR TITLE
Add more Canon RF lenses

### PR DESCRIPTION
**Pull request by kmilos**
Sync tables w/ 8 missing from exiftool

LensType entries have been re-sorted according to RFLensType value; the number sequence in the comments doesn't have any meaning, as there are gaps to be filled... [Exiftool also re-sequences these](https://github.com/exiftool/exiftool/commit/15e88b0fcedb10f6348610b562725e4dfcf2b3e1#diff-d7ae93c137f0b89f8b7bc35026653055cf6181c69d45337b86b4fae16757908c).

### CODE REVIEW

**Comment by postscript-dev** [comment]
@kmilos:
Re-ordering the `Exif.CanonCs.LensType` values is a good idea, perhaps this could be applied elsewhere in the future. One query though, do we still need to provide a test image for each new lens that we are adding?

**Comment by kmilos** [comment]
Thanks for taking a look. AFAICT, we don't have a single Canon RF test case yet...

**Comment by postscript-dev** [comment]
@piponazo: I was hoping for a little advice on this PR so that we can move forward.

While `Exif.CanonCs.LensType` has lots of test files, `Exif.CanonFi.RFLensType` doesn't have *any* testing to prove that the lenses are correct. There are a couple of Canon files in Exiv2 that we could make use of but the other 20+ entries would still be untested. This new PR adds some extra lenses to `Exif.CanonFi.RFLensType` but doesn't add any new test files.

I was wondering how to proceed. Do we still require that all changes to the translated lens values need a test file?

**Comment by piponazo** [comment]
Hi folks. Unfortunately, that part of the code base (lenses) is the one I know less about. I have never seen a test regarding this part of the code and I do not know if it is feasible to add something to test it out.

For me it would be fine to add these changes without additional testing effort since it is not breaking any existing test. We could also create a new issue to investigate how to properly test that area of the code if it something you think we should explore.

File: src/canonmn_int.cpp [hunk 1]
```diff
@@ -1854,34 +1854,37 @@ constexpr TagDetails canonCsLensType[] = {{1, "Canon EF 50mm f/1.8"},
...
-                                          {61182, "Canon RF 70-200mm F2.8L IS USM"},                // 18
-                                          {61182, "Canon RF 600mm F11 IS STM"},                     // 19
-                                          {61182, "Canon RF 600mm F11 IS STM + RF1.4x"},            // 20
-                                          {61182, "Canon RF 600mm F11 IS STM + RF2x"},              // 21
-                                          {61182, "Canon RF 800mm F11 IS STM"},                     // 22
-                                          {61182, "Canon RF 800mm F11 IS STM + RF1.4x"},            // 23
-                                          {61182, "Canon RF 800mm F11 IS STM + RF2x"},              // 24
-                                          {61182, "Canon RF 24-105mm F4-7.1 IS STM"},               // 25
-                                          {61182, "Canon RF 100-500mm F4.5-7.1L IS USM"},           // 26
-                                          {61182, "Canon RF 100-500mm F4.5-7.1L IS USM + RF2x"},    // 27
```
Comment (hassec) [inline src/canonmn_int.cpp:1884]: I'd prefer if you could not remove these comments, it introduces an unnecessary diff.  
And other duplicate values still have them in this array, so let's just leave them and stay consistent.

**Review by hassec — DISMISSED** [review]

**Comment by hassec** [comment]
I think it would be a nice idea to investigate if we can extend `test_canon_lenses.py` to also test the lenses defined in `canonRFLensType`  

Not necessarily for this PR though,  but probably worth creating an issue to follow up.

**Comment by postscript-dev** [comment]
@piponazo: Thanks for replying.

In the past, we asked the user for a test file taken with the lens which was then added as a new Python test (e.g., #1784).

@hassec: As long as you have free time, I am happy to leave this for you to review.

**Comment by hassec** [comment]
I agree with @postscript-dev that it would be nice to introduce these kinds of test images if possible.

File: src/canonmn_int.cpp (hunk 1 above)
Comment (kmilos) [reply src/canonmn_int.cpp:1884]: I.e. maintaining the comments for RF lenses and re-sequencing every time a lens is added to match exiftool will introduce unnecessary diffs in the future. Adding in sequence will make syncing w/ `RFLensType` table that has a clear sorting order by ID.

Comment (hassec) [reply src/canonmn_int.cpp:1884]: Well IMHO the sorting here is not obvious because the RF ID doesn't exist in this list as it doesn't have any meaning. In the CsType list they all have only one ID (61182).     
Thus, it's only a matter of time until someone adds a lens to this table in what you would call the "wrong order", which is why I'd just from the start not worry about the order and ignore it. As I said, the order is meaningless here anyway.   

If you want to make this nicer, I'd argue it's better to have the value of `61182` directly point to the `canonRFLensType` list which would avoid the duplication.

Comment (kmilos) [reply src/canonmn_int.cpp:1884]: exiftool keeps two lists and the RF lenses always sorted in the ID order in both (don't know if it's manual or by a script). We could simply put a comment above the first 61182 lens: "please keep in sync w/ sorted RFLensType list below"

> If you want to make this nicer, I'd argue it's better to have the value of `61182` directly point to the `canonRFLensType` list which would avoid the duplication.

Yes, please, but it's not obvious to me how to do this ATM.

Comment (hassec) [reply src/canonmn_int.cpp:1884]: adding `"please keep in sync w/ sorted RFLensType list below"` would be a compromise I can live with :wink:

Comment (kmilos) [reply src/canonmn_int.cpp:1884]: Will do, just have to get around to it...

File: src/canonmn_int.cpp [hunk 2]
```diff
@@ -2260,7 +2266,15 @@ constexpr TagDetails canonRFLensType[] = {{0, N_("n/a")},
                                           {276, N_("Canon RF 100-500mm F4.5-7.1L IS USM + RF1.4x")},
                                           {277, N_("Canon RF 100-500mm F4.5-7.1L IS USM + RF2x")},
                                           {278, N_("Canon RF 70-200mm F4L IS USM")},
-                                          {280, N_("Canon RF 50mm F1.8 STM")}};
+                                          {280, N_("Canon RF 50mm F1.8 STM")},
```
Comment (postscript-dev) [inline src/canonmn_int.cpp:2269]: @kmilos:
It is only a suggestion but perhaps you could remove the `N_( )` translation functions from all the `canonRFLensType` entries. As far as I understand it, these strings do not need translating.

Removing the functions is consistent with the `canonCsLensType` array and the other makernotes' lens values.

**Comment by kmilos** [comment]
Thanks - although I didn't introduce those originally, I have nothing against removing them. Mind you, the gettext N_() macro is a noop and is not incorrect here (i.e. actually explicitly prevents any extraction for translation).

File: src/canonmn_int.cpp (hunk 2 above)
Comment (kmilos) [reply src/canonmn_int.cpp:2269]: Thanks - although I didn't introduce those originally, I have nothing against removing them. Mind you, the gettext N_() macro is a noop and is not incorrect here (i.e. actually explicitly prevents any extraction for translation).

Comment (postscript-dev) [reply src/canonmn_int.cpp:2269]: @kmilos:
> Thanks - although I didn't introduce those originally, I have nothing against removing them. Mind you, the gettext N_() macro is a noop and is not incorrect here (i.e. actually explicitly prevents any extraction for translation).


I think that using `N_( )` declares which text *is* made available for translation (assuming that you are compiling with `EXIV2_ENABLE_NLS` set to `ON`). The `N_( )` definition can be found in [src/i18n.h](https://github.com/Exiv2/exiv2/blob/7ebf2a184e28bf638a39aa30ed3784c5035b668a/src/i18n.h).

For example, the results can be see with the description string:

https://github.com/Exiv2/exiv2/blob/7ebf2a184e28bf638a39aa30ed3784c5035b668a/src/canonmn_int.cpp#L1967-L1969

which is translated in:

https://github.com/Exiv2/exiv2/blob/7ebf2a184e28bf638a39aa30ed3784c5035b668a/po/de.po#L627-L633

If `N_( )` is removed from the RF lenses then this prevents them from being passed to the translators. Including this is only a suggestion though and it is completely fine to leave it out.

As an aside, we probably have other existing text where the same idea could be applied but reviewing every string is a **HUGE** job! I have spotted a couple of small changes but have no plans to review everything.

Comment (kmilos) [reply src/canonmn_int.cpp:2269]: Btw, those examples you gave do not come from the same source file: tranlatable is extracted from `src/canonmn.cpp:1230`, while the string you showed is from `canonmn_int.cpp`.

Comment (postscript-dev) [reply src/canonmn_int.cpp:2269]: @kmilos:

> N_() is no-op, it is definitely not extracted for translation, unlike _(), which is.

Looking again at this, I can see you point. I was misled by the `.po` files containing the correct translations strings. Although I don't use it myself, as far as I know, the translation system does work.

I can see that we do use some `_( )` but most references seem to be `N_( )`. In the past I have always used whatever function was already in the code.

Comment (postscript-dev) [reply src/canonmn_int.cpp:2269]: @kmilos:
No problem, thanks for checking.

Comment (kmilos) [reply src/canonmn_int.cpp]: Thanks for making me do it ;)

**Comment by kmilos** [comment]
Synced w/ exiftool 12.45. Please review again @hassec

**Comment by kmilos** [comment]
I have now also padded all the model IDs to 32b and re-sorted, and added more lenses up to exiftool 12.49

**Review by postscript-dev — DISMISSED** [review]
You could add `Exif.CanonFi.RFLensType` to https://github.com/Exiv2/exiv2/blob/1f5bc3b8a558f2d1935a57d469a492fbc937c912/src/easyaccess.cpp#L232-L239 However it is unlikely to be used as `Exif.CanonCs.LensType` is likely to exist and would have priority. Adding the tag is only to keep the array complete.

**Comment by kmilos** [comment]
Thanks, but I think @hassec idea of redirecting this automatically when Exif.CanonCs.LensType is 68112 is the best way to go instead. Haven't thought about how to implement it yet, and would prefer we merge this as is, and open a separate FR.

**Comment by postscript-dev** [comment]
@kmilos 
> Thanks, but I think @hassec idea of redirecting this automatically when Exif.CanonCs.LensType is 61182 is the best way to go instead. Haven't thought about how to implement it yet, and would prefer we merge this as is, and open a separate FR.

I was thinking of suggesting the same thing. I had two other thoughts about the easyaccess API that may be of interest for this PR:

1. In https://github.com/Exiv2/exiv2/blob/1f5bc3b8a558f2d1935a57d469a492fbc937c912/src/easyaccess.cpp#L232-L239 I don't know why `Exif.CanonCs.LensType` is listed before `Exif.Photo.LensModel` (commit 9ac47a73330107a8ec4ffe6f2713855686e976c2) but it was changed before the exiv2 config file was added. If the order of the tags were reversed, then for modern cameras/lenses, the user would only get one answer instead of a list of alternatives.

2. You could add `Exif.Canon.ModelID` to https://github.com/Exiv2/exiv2/blob/1f5bc3b8a558f2d1935a57d469a492fbc937c912/src/easyaccess.cpp#L310-L318 However, using your `pr_2216.exv` file, we can see that the output is not exactly the same:

```shell
$ exiv2 -Pxkycvt -K Exif.Image.Model -K Exif.Canon.ModelID pr_2216.exv
0x0110 Exif.Image.Model                             Ascii      12  Canon EOS R  Canon EOS R
0x0010 Exif.Canon.ModelID                           Long        1  2147484708  EOS R
```

Some camera manufacturer's include the make in `Exif.Image.Model`, some don't (e.g., Sony) and others seem inconsistent (e.g., Nikon).

**Comment by kmilos** [comment]
I really, really, really, don't want to go down the lens recognition rabbit hole further.

This PR was just about bringing the lists up to date wrt exiftool. If something that simple is not mergeable for several months now, I'll happily drop it in favour of other PRs.

**Comment by postscript-dev** [comment]
> I really, really, really, don't want to go down the lens recognition rabbit hole further.

> This PR was just about bringing the lists up to date wrt exiftool. If something that simple is not mergeable for several months now, I'll happily drop it in favour of other PRs.

I was only offering suggestions but they can be dealt with in a different PR. I didn't know your intentions. I will approve this PR.

**Comment by kmilos** [comment]
Thanks @postscript-dev Before you do, I'll just add the exiftool link @hassec requested.

File: src/canonmn_int.cpp (hunk 1 above)
Comment (kmilos) [reply src/canonmn_int.cpp:1936]: Done

**Comment by postscript-dev** [comment]
@kmilos 
Can you request a review from me again, as I am having a problem triggering the process.

**Comment by kmilos** [comment]
@postscript-dev Ok now? You should be able to go to "Files changed" and just finish it there at any time, no?

**Review by postscript-dev — APPROVED** [review]

**Comment by kmilos** [comment]
Thanks you Peter! 🥇
